### DISCUSSION TITLE
General beginner questions - yocto for nvidia jetsons

### DISCUSSION THREAD

**Discussion by hutcruchi** [OE4T/meta-tegra#1312] [Q&A] [answered]
Hi,
I apologies in advance if this is not the suitable place for asking my beginner's questions.
I hope to get some insights and "first steps" guidance from the experienced users here.

I have little experience working with several Nvidia Jetson systems (SoMs and devkits). 
My main goal is to reduce boot time as much as possible and this led me to my current task - create my own minimal rootfs.
The main purpose of using Yocto is educational. I have no prior experience with Yocto, and before I enter this world (and steep learning curve) I would like to ask a few questions just to make sure this is the right path for me:

1. Do you even recommend using Yocto for building minimal custom rootfs for Nvidia jetsons? (using their BSP with my new Yocto maid minimal rootFS)

2. From your experience, are there any known limitations/incompatibilities when using Yocto to build rootfs/custom distribution for Nvidia Jetsons (Xavier NX, Orin NX, JP5.1.1), maybe some of their content (apps,libraries,features) just wont work?, maybe they supply some compiled products (without source) suitable only for their environment?

3. Basic understanding of rootFS:
I don't understand why my rootFS "has to be attached to specific hardware". I'll explain:
From my understanding, the kernel ships with statically built in drivers and kernel modules (in and out of tree). These are the drivers that I need in order for the system to work with my specific hardware. So why can't I use the same rootFS for each hardware (in case of need changing dtbs ,bootloader etc), and just compile and use the drivers needed for my specific hardware? , in other words - what does the rootfs contains that is specific to certain hardware.

4. Last general question:
Lets say I used some quick guide doc to build my first "vanila" custom rootFS.  What are the first steps needed in order to match it to the specific hardware (in our case the jetson). Which "check list" I have to fill in order to declare that this rootfs is suitable for the specific jetson. 

Thank you very much

**Comment by dwalkes** [comment]
> I apologies in advance if this is not the suitable place for asking my beginner's questions.

Welcome, and this is a fine place.  You can also use the [gitter-im community](https://github.com/OE4T/meta-tegra/wiki) for more interactive discussion if you prefer.

> Do you even recommend using Yocto for building minimal custom rootfs for Nvidia jetsons? (using their BSP with my new Yocto maid minimal rootFS)

Yes I do recommend this, but I guess this group will probably be biased ;)  The benefit of using Yocto is the ability to customize exactly what you want in the rootfs and to build it in a reliable, reproducible, and source tracked/controlled manner.

> From your experience, are there any known limitations/incompatibilities when using Yocto to build rootfs/custom distribution for Nvidia Jetsons (Xavier NX, Orin NX, JP5.1.1), maybe some of their content (apps,libraries,features) just wont work?

I don't think there's anything that won't work AFAIK.  There may be things which aren't working because nobody is focused on/using them in their application.  In this case you'd want to open an issue with the specific problem you have.

> maybe they supply some compiled products (without source) suitable only for their environment?

There are lots of components delivered in binary form without source, and in some cases we've needed to do some workarounds to decouple them from the Jetson specific rootfs but I'm not aware of any specific outstanding issues where it's not possible to use some feature for this reason.  If this was the case we could ask NVIDIA to help us fix this.  They are generally supportive of this project and have made changes in the past based on our requests.

> I don't understand why my rootFS "has to be attached to specific hardware". I'll explain:
> in other words - what does the rootfs contains that is specific to certain hardware.

In NVIDIA's case it's closed source binary libraries which make assumptions about specific kernel/bootloader configurations.  You could likely make a rootfs which isn't specific to NVIDIA but in that case most of the reason you'd want to use the hardware in the first place (accelerated video encode/decode, cuda, etc) wouldn't be available, which would defeat the purpose of using the hardware.

> What are the first steps needed in order to match it to the specific hardware (in our case the jetson). Which "check list" I have to fill in order to declare that this rootfs is suitable for the specific jetson.

You'd probably want to make this checklist yourself based on how you need to use the hardware for your specific application.  Presumably there are some hardware features you will need for your application, and you'll want to make sure those are working properly with your rootfs.  We have some test cases/scripts we use to verify some of the possible software components which you can find referenced in the release and validation sheet linked [here](https://github.com/OE4T/meta-tegra/wiki/OE4T-Contributor-Guide).  NVIDIA also has some checklists in their bring up guides which might be useful.  But ultimately I think you'd want to start with your specific application rather than try to verify every possible function is working properly.

**Reply by hutcruchi** [reply]
First of all, thank you very much for your reply. Much appreciate!

Secondly, I apologies for kinda return the same question in other words. I didn't fully realized the practical tasks involved in creating custom distro and I'm a little bit confused.

For the sake of my question, our use case is comparing between Nvidia's official BSP and a Yocto custom distro. I wish to understand the required amount of work for making our custom distro fully functional (given our needs) on a specific jetson hardware.

1. Toolchain-
Nvidia doesn't own the toolchain, hence in the worst case I can build and use it in our Yocto distro. 
Bottom line: no custom work needed here. 
2. Kernel , Drivers, DT-
Nvidia doesn't own the kernel sources, make system, statically in tree drivers. In the worst case they own some out of tree drivers which are supplied by them as modules and we can use them in our Yocto distro. Maybe we have to install some packages by ourselves but that's it. 
Bottom line: no custom work needed here.
3. Bootloader, UEFI- 
If we can use their bootloader with our distro its a win. If not than maybe we have to use another bootloader (like uboot) and make the adjustments. If their bootloader's source code is available than maybe we can modify its code. 
Bottom line: in many cases - custom work is needed.
4. Applications, Libraries:
same as (3)

So, from these points I mentioned I don't understand why it involves a lot of work to create a layer enabling our custom distro working with the vendor hardware. For most of the tasks I mentioned the bottom line is that no custom work is needed.

I can print a list of all packages included in the vendors BSP and install them in my distro (the ones I want)
I can print a list of all drivers included in the vendors BSP (modules and static) and install them in my distro (the ones I want)
 so what I missing here? I know my understanding is not complete/correct. There is a lot of work being done that somehow solved by using the correct layers/receipts and indicating the correct machine because other people already worked hard for me. 
But what they did exactly? and what they haven't done yet that I need to do by myself?


Thanks again

**Reply by dwalkes** [reply]
> First of all, thank you very much for your reply. Much appreciate!

You are welcome

>  I didn't fully realized the practical tasks involved in creating custom distro and I'm a little bit confused.

You don't necessarily need to create a custom distro from scratch, you can use the example in https://github.com/OE4T/tegra-demo-distro and [poky](https://www.yoctoproject.org/software-item/poky/) as a starting point.

As long as you are using one of the supported hardware configurations in meta-tegra this will take care of most of your items 1-4 above.

Even if your hardware isn't directly supported you will likely be able to leverage a similar hardware configuration for your custom machine definition.

> I can print a list of all packages included in the vendors BSP and install them in my distro (the ones I want)

Assuming these packages are [also in the yocto layer index](https://layers.openembedded.org/layerindex/branch/master/recipes/) you likely don't need to do much work to include them in your custom distro, possibly just [IMAGE_INSTALL:append](https://docs.yoctoproject.org/dev-manual/customizing-images.html) with whatever packages you wish to include.

>  There is a lot of work being done that somehow solved by using the correct layers/receipts and indicating the correct machine because other people already worked hard for me. But what they did exactly? and what they haven't done yet that I need to do by myself?

I think the easiest way to understand this is by attempting to build a reference distribution and run it, using the instructions in https://github.com/OE4T/tegra-demo-distro, and comparing this to your requirements.  In terms of what you need to do yourself, this will require more information about what you are trying to do.  If you are trying to boot a development board (or similar carrier board based on development board design) already supported by meta-tegra and add some packages already supported by existing yocto layers there's probably not a lot of work you'll need to do yourself.  It's hard to speculate based on what you've written here thus far.

**Reply by hutcruchi** [reply]
Thanks again for your detailed answer.
I have few **last** questions by your permission:

1. Does the tegra-demo-distro considered as "light" or "minimal" distro? I'm asking to know if I can use it as a starting point distro and just add things on top of it for my custom hardware systems when my goal is a thin system booting fast so I don't want it to already contain unnecessary stuff.

2. General question:
If I use this distro, edit its applications, properties, rootfs,etc and now I want to use it on a different board (Xilinx for example). Should I change the hardware related layers only or should I start over from a different minimal reference distro?
My goal is to use the same distro on different hardwares.

3. General question:
When a custom distro is being made, does it based on an existing known distro? for exmaple I want to create a new minimal distro. Does it based on an Ubuntu or Centos or other distro of my choice? I'm asking this because probably we don't build entire distro from scratch. it must be based on something, otherwise its a lot of work

**Comment by dwalkes** [comment, marked as answer]
> Does the tegra-demo-distro considered as "light" or "minimal" distro

There are several images to choose from depending on what you'd like included.  core-image-minimal is the smallest but may not be very useful because it does not contain much of the Nvidia content you'll likely need for your application. See https://docs.yoctoproject.org/ref-manual/images.html#:~:text=core%2Dimage%2Dminimal%20%3A%20A,allowing%20a%20device%20to%20boot. For an overview of the built in image types.

> If I use this distro, edit its applications, properties, rootfs,etc and now I want to use it on a different board (Xilinx for example). Should I change the hardware related layers only or should I start over from a different minimal reference distro?

Theoretically it should be possible to use with different hardware types.  For Xilinx specifically I'm not knowledgeable enough about their Yocto BSP to know whether this is feasible.  Typically you'd put any of your custom content in your own layer and then share this layer between multiple projects, but perhaps not share the full build environment between both.  Note that you can define package lists in your shared layer which could then be used in multiple hardware targets.

> When a custom distro is being made, does it based on an existing known distro? for exmaple I want to create a new minimal distro. Does it based on an Ubuntu or Centos or other distro of my choice

It's based on the poky images linked above.  If you are looking for a way to build a Debian distribution you could explore a project like https://github.com/siemens/isar instead, although I'm not familiar enough with that myself to understand how it could integrate with something like meta-tegra.

**Reply by hutcruchi** [reply]
Thank you for your answers. Much appreciate !
